corehq/form_processor/tasks.py

```diff
@@ -37,19 +39,24 @@ def reprocess_archive_stubs():
     # Check for archive stubs
     from corehq.form_processor.interfaces.dbaccessors import FormAccessors
     from couchforms.models import UnfinishedArchiveStub
-    stubs = UnfinishedArchiveStub.objects.filter()
+    stubs = UnfinishedArchiveStub.objects.filter(attempts__lt=3)
```

Comment:
What do you think of adding a [partial index](https://www.postgresql.org/docs/9.6/indexes-partial.html) on `attempts < 3` so this lookup remains fast? Without the index this query will always scan all stubs with too many attempts, which will eventually get slow when too many stubs with >= 3 attempts accumulate.

Another possible solution would be to delete stubs with too many attempts. This might be better, but I'm not sure if we need to keep a record of failed attempts.

Reply:
That's interesting. I am also not totally sure what to do about persistent failures, and PR very much kicks that can down the road. These failures I think represent actual bugs in our code that are hit by a shockingly small fraction of a simple size that's in the millions, as far as I can tell. (There appear to be at most 22 errors like this since we introduced this retry behavior a little over a year ago.) In the case of a short-lived bug though, it would be nice to reset attempts and retry everything after its fixed, which we'd have to do manually or write into this code when it comes up. For now I went with keeping them around so that they're there to pull from if / when we need that feature.

Reply:
That's interesting. I am also not totally sure what to do about persistent failures, and PR very much kicks that can down the road. These failures I think represent actual bugs in our code that are hit by a shockingly small fraction of a simple size that's in the millions, as far as I can tell. (There appear to be at most 22 errors like this since we introduced this retry behavior a little over a year ago.) In the case of a short-lived bug though, it would be nice to reset attempts and retry everything after its fixed, which we'd have to do manually or write into this code when it comes up. For now I went with keeping them around so that they're there to pull from if / when we need that feature.

As for the index, I'm not sure it's necessary but it also is very unlikely to hurt. As far as I can tell django doesn't nicely support partial indices until Django 2.x. What do you think about just putting a normal index on attempts? I can't imagine at the scale we're talking about it would make any kind of performance difference, but that might say more about my imagination than about me.

Reply:
Here's an example of where we create partial indexes: https://github.com/dimagi/commcare-hq/blob/3051be40a0b4a19520383471dd82ec48474745a2/corehq/blobs/models.py#L74-L78

Though I agree that I don't think it's worth it unless we end up with lots of failed attempt rows in which case we should likely delete them or fix the issues.

Reply:
I suppose it's reasonable to wait to add the partial index until it becomes a problem. Sounds like that may be a long time, given the rate at which these build up.